### ISSUE TITLE
Search from each selection

### ISSUE THREAD

**Issue by the-mikedavis** [helix-editor/helix#5672]
The current search (`/`, `n` and `N`) only searches from the primary cursor. In select mode (`v/`) this can be used to add matches of the search pattern as new cursors.

It would also be useful to be able to search from every cursor. Instead of adding new selections for the next match from the primary cursor, this search would replace or extend the current selection(s) (depending on normal or select mode).

This behavior would be the same as textobjects like `]p` or `]f`. This could be used together with #1383 to implement regex-based textobjects like searching for numbers, dates, version control markers (https://github.com/helix-editor/helix/issues/4647), etc.. And I think it could be useful on its own in some multi-cursor scenarios and some single-cursor scenarios like #5668.

There might be some overlap with this and https://github.com/helix-editor/helix/pull/4635 with how search matches would be computed.

**Comment by fennewald** [comment]
Ran into a case were running a macro with multiple cursors will cause the macro to behave differently for primary vs the others. Would definitely appreciate this!

**Comment by xxxserxxx** [comment]
I want to comment here a work-around, for anyone (e.g., future me) searching for how to accomplish this (until such time as something better is implemented):
```
v<Alt-/>[search term]<Alt-->
```
and then trim the selection as necessary.

**Comment by sibouras** [comment]
you could `/ccc<Enter>C/fff` then `<alt>-`

**Comment by the-mikedavis** [comment]
`A-/` has never been bound in the keymap - you can search for it with `git log -S 'A-/'` and `git log -S "alt!('/"`.

For that use-case we could respect the mode in `/`/`?` so that it behaves like `n`/`N` in select mode. Currently it's hard-coded to move the primary cursor rather than add a cursor like you would with `n`/`N` in select mode. (That's why @sibouras's workaround works - searching in normal mode moves the primary cursor and leaves the rest unchanged.) Then you could merge selections with `A-minus` to select the whole range.

**Comment by rcorre** [comment]
> but how do you get to fff extending your current selection?

`vgw` works, using the new `extend_to_word` (not saying that fixes the issue, but just another workaround to be aware of).

**Comment by xxxserxxx** [comment]
> `A-/` has never been bound in the keymap - you can search for it with `git log -S 'A-/'` and `git log -S "alt!('/"`.

Yeah, I don't remember what I meant to write; I had A-/ bound to something, or I just typed it in from a faulty memory. The entire work-around was so klugy I rarely used it. @sibouras' solution does the same thing and is cleaner.

Cognitively, being in visual mode and having selected some text, and having searching forward (or backward) causing the selection to be lost -- I'm trying to understand why that is the current behavior. Is that the intuitive behavior for someone, or is it some technical difficulty?

**Comment by the-mikedavis** [comment]
I believe it's just an oversight. Currently select mode has no effect in `search`/`rsearch` but I don't think that's intentional

**Comment by sibouras** [comment]
it would make more sense for `gw` to respect mode also like `/` `n` `N`, and you can get the old behavior by `merge_selections`

**Comment by the-mikedavis** [comment]
That isn't about respecting mode. `search`/`rsearch` handle select mode differently than all other commands by adding selections rather than extending them. `gw`'s behavior is the way select mode should generally work.

**Comment by sibouras** [comment]
ah i see, it would be nice though when in select mode `gw` selects a new word, it would make swapping words easier eg.

`vegw` then `rotate_selection_content_forward` instead of having to type or regex find the word, and we can have the old behavior by `merge_selections`

**Comment by thomasschafer** [comment]
If we're thinking that `/` should respect mode, presumably `n` should do the same, which would conflict with the current behaviour of `n` in visual mode adding a selection at the next occurrence of the search term - would we want a new keymap for this existing behaviour?

I want to try and get a PR up to fix this behaviour but just want to confirm the desired behaviour before I get started.

**Comment by pascalkuthe** [comment]
@the-mikedavis is already testing an Implementation of this that we discussed that matches what we want so starting from scratch with a new PR probably doesn't make sense

**Comment by thomasschafer** [comment]
Ah, thanks for letting me know! I wasn't aware this was being worked on.

As an aside, I also noticed a slight bit of inconsistency in that using `gw` to jump ahead maintains the selection when in visual mode, but using `gw` to jump to a word within the selection does not. Assuming this is unintentional, and the expected behaviour is more like that of Vim where the anchor would remain fixed and the selection cursor would just move backwards, I'd be interested to know if this has also been fixed as part of the above implementation or if this is something to be looked at separately? Happy to have a crack at it myself if the latter.

https://github.com/helix-editor/helix/assets/54135831/0c650f94-6dd0-4bdf-9ae9-08e31a630b22

**Comment by the-mikedavis** [comment]
That's intentional: https://github.com/helix-editor/helix/pull/8875#discussion_r1428152259

**Comment by antislava** [comment]
Inspired by @rcorre 's workaround, I just tried this:

```toml
[keys.select]
'/' = ["search", "merge_selections"]
'?' = ["rsearch", "merge_selections"]
```

Seems to work well with `Alt-minus`, e.g. when I want to select the remaining part of a long markdown code block. (Could come with different shortcuts instead of '/' and '?', but haven't noticed any conflicts so far.) What do you think?

**Comment by antislava** [comment]
helix 25.01.1 (b42e1d20). (I don't think I have anything fancy in my config that could affect this...)

**Comment by xxxserxxx** [comment]
Weird. Me too.

Just to clarify, after adding this, you can go into visual mode, select some text, press `/` and search for some text, and it'll include everything between your visual-selected start and the search results?

**Comment by antislava** [comment]
@xxxserxxx After entering selection mode (with `v`), searching for arbitrary patterns (not necessarily the same) will add a new cursor on each found pattern instance. Pressing `Alt-minus` creates a continuous selection between the initial and the last cursor. (I don't mind pressing `Alt--` at all and the "cursor collection" before that can be useful on its own. To me, this feels even superior to the traditional vim behaviour, actually.)
